json_rpc: while parsing the json read from unix socket I receive a parser error

This is what is reported by the CI [here](https://github.com/vincenzopalazzo/cln4zig/runs/7550429910?check_suite_focus=true) and the same json parsing is also tested [here](https://github.com/vincenzopalazzo/cln4zig/blob/master/src/json_rpc/rpc_two.zig#L76-L91) but looks like if the string contains trash at the end the json parser will panic when finding a wrong token at the end of JSON.

```
| zig test src/*.zig
cln_1  | 1/2 test "try to call core lightning Unix RPC method"... {"jsonrpc":"2.0","id":"1","result":{"id":"02046bdac40419c3c01aa8f43a6f0b0724464408f11e39df954fa5ce44d1e72565","alias":"YELLOWSUCKER_TES_ONE","color":"03ad98","num_peers":0,"num_pending_channels":0,"num_active_channels":0,"num_inactive_channels":0,"address":[],"binding":[{"type":"ipv4","address":"127.0.0.1","port":19735}],"version":"0.11.2","blockheight":0,"network":"regtest","msatoshi_fees_collected":0,"fees_collected_msat":"0msat","lightning-dir":"/workdir/lightning_dir_one/regtest","warning_lightningd_sync":"Still loading latest blocks from bitcoind.","our_features":{"init":"080269a2","node":"800000080269a2","channel":"","invoice":"02000000024100"}} }FAIL (InvalidTopLevelTrailing)
cln_1  | /usr/local/zig-linux-x86_64-0.10.0-dev.3315+1a16b7214/lib/std/json.zig:403:21: 0x2353e2 in std.json.StreamingParser.transition (test)
cln_1  |                     return error.InvalidTopLevelTrailing;
cln_1  |                     ^
cln_1  | /usr/local/zig-linux-x86_64-0.10.0-dev.3315+1a16b7214/lib/std/json.zig:325:13: 0x2299a6 in std.json.StreamingParser.feed (test)
cln_1  |         if (try p.transition(c, token1)) {
cln_1  |             ^
cln_1  | /usr/local/zig-linux-x86_64-0.10.0-dev.3315+1a16b7214/lib/std/json.zig:1120:13: 0x2280ea in std.json.TokenStream.next (test)
cln_1  |             try self.parser.feed(self.slice[self.i], &t1, &t2);
cln_1  |             ^
cln_1  | /usr/local/zig-linux-x86_64-0.10.0-dev.3315+1a16b7214/lib/std/json.zig:1817:16: 0x22758e in std.json.Parser.parse (test)
cln_1  |         while (try s.next()) |token| {
cln_1  |                ^
cln_1  | /workdir/code/src/json_rpc/rpc_two.zig:70:25: 0x220eda in json_rpc.rpc_two.JSONRPC.read (test)
cln_1  |         var json_tree = try parser.parse(string);
cln_1  |                         ^
cln_1  | /workdir/code/src/json_rpc/rpc_two.zig:21:9: 0x22062f in json_rpc.rpc_two.JSONRPC.call (test)
cln_1  |         return self.read(&arena);
cln_1  |         ^
cln_1  | /workdir/code/src/main.zig:26:16: 0x21e4ff in CLNUnix.call (test)
cln_1  |         return try self.socket.call("1", method, payload);
cln_1  |                ^
cln_1  | /workdir/code/src/main.zig:45:9: 0x21d565 in test "try to call core lightning Unix RPC method" (test)
cln_1  |     _ = try client.call("getinfo", json.ObjectMap.init(allocator.allocator()));
cln_1  |         ^
cln_1  | 2/2 json_rpc.rpc_two.test "check parser function"... OK
cln_1  | 1 passed; 0 skipped; 1 failed.
cln_1  | error: the following test command failed with exit code 1:
cln_1  | src/zig-cache/o/5ab7118f61af2118347843d2a5a1de5d/test /usr/local/zig-linux-x86_64-0.10.0-dev.3315+1a16b7214/zig
cln_1  | make: *** [Makefile:10: check] Error 1
```

--- a/src/json_rpc/rpc_two.zig
+++ b/src/json_rpc/rpc_two.zig
@@ -24,6 +24,7 @@ pub const JSONRPC = struct {
     fn build_message(self: *Self, arena: *std.heap.ArenaAllocator, id: []const u8, method: []const u8, payload: json.ObjectMap) ![]const u8 {
         const allocator = arena.allocator();
         var request_tree = json.ObjectMap.init(allocator);
+        defer request_tree.deinit();
         try request_tree.put("id", std.json.Value{ .String = id });
         try request_tree.put("jsonrpc", std.json.Value{ .String = self.version });
         try request_tree.put("method", std.json.Value{ .String = method });
@@ -71,6 +72,6 @@ test "check parser function" {
 pub const ResponseObj = struct {
     id: ?[]const u8,
     jsonrpc: []const u8,
-    //errors: ?json.ObjectMap,
+    errors: ?json.ObjectMap,
     result: ?json.ObjectMap,
 };

--- a/src/json_rpc/rpc_two.zig
+++ b/src/json_rpc/rpc_two.zig
@@ -15,7 +15,7 @@ pub const JSONRPC = struct {
 
     pub fn call(self: *Self, id: []const u8, method: []const u8, payload: json.ObjectMap) !json.Value {
         var arena = std.heap.ArenaAllocator.init(std.heap.page_allocator);
-        //defer arena.deinit();
+        defer arena.deinit();
         var to_stream = try self.build_message(&arena, id, method, payload);
         _ = try self.stream.write(to_stream);
         return self.read(&arena);
@@ -38,35 +38,12 @@ pub const JSONRPC = struct {
         var reader = self.stream.reader();
         var buffer = std.ArrayList(u8).init(arena.allocator());
         defer buffer.deinit();
-        while (true) {
-            const size_buffer = 64;
-            var tmp_buffer: [64]u8 = undefined;
-            var byte_read = reader.read(&tmp_buffer) catch unreachable;
-            // FIXME: if the read byte is less of the buffer we got stuck in the read
-            // if (byte_read == 0) {
-            //     std.debug.print("{s}\n", .{"end buffer"});
-            //     break;
-            // }
-            buffer.appendSlice(&tmp_buffer) catch unreachable;
-            if (byte_read < size_buffer) {
-                try buffer.append(tmp_buffer[byte_read - 1]);
-                break;
-            }
-        }
-        var string = buffer.allocatedSlice();
-        var clean_string = std.ArrayList(u8).init(arena.allocator());
-        for (string) |char, index| {
-            if (char == '\n' and (index > 0 and string[index - 1] == '\n')) {
-                break;
-            } else if (char == '\n') {
-                continue;
-            }
-            try clean_string.append(char);
-        }
-        string = clean_string.allocatedSlice();
+        try reader.readUntilDelimiterArrayList(&buffer, '\n', 40000);
         var parser = std.json.Parser.init(arena.allocator(), false);
         defer parser.deinit();
-        std.debug.print("{s}", .{string});
+        var string = buffer.toOwnedSlice();
+        std.debug.print("{s}\n", .{string});
+        std.debug.assert(std.json.validate(string) == true);
         var json_tree = try parser.parse(string);
         defer json_tree.deinit();
         return json_tree.root;
@@ -81,11 +58,12 @@ test "check parser function" {
 
     var string =
         \\ {"jsonrpc":"2.0","id":"1","result":{"id":"03b39d1ddf13ce486de74e9e44e0538f960401a9ec75534ba9cfe4100d65426880",
-        \\ "alias":"SLICKERGOPHER-testnet","color":"02bf81","num_peers":17,"num_pending_channels":1,"num_active_channels":16,
-        \\ "num_inactive_channels":0,"address":[{"type":"torv3","address":"iopzsfi3pbtrimnovncctjthmot7nb6pvh2swr6pqvizmisy4ts62lid.onion","port":19735}],
-        \\ "binding":[{"type":"ipv4","address":"127.0.0.1","port":9735}],"version":"0.10.2","blockheight":2282149,
-        \\ "network":"testnet","msatoshi_fees_collected":30021,"fees_collected_msat":"30021msat",
-        \\ "lightning-dir":"/media/vincent/VincentSSD/.lightning/testnet"} }
+        \\ "alias":"SLICKERGOPHER-testnet","color":"02bf81","num_peers":20,"num_pending_channels":1,"num_active_channels":22,
+        \\ "num_inactive_channels":0,"address":[{"type":"torv3","address":
+        \\ "iopzsfi3pbtrimnovncctjthmot7nb6pvh2swr6pqvizmisy4ts62lid.onion","port":19735}],"binding":[{"type":"ipv4",
+        \\ "address":"127.0.0.1","port":19735}],"version":"v0.12.0rc3-4-g3ce2673-modded","blockheight":2344351,
+        \\ "network":"testnet","fees_collected_msat":30021,"lightning-dir":"/media/vincent/VincentSSD/.lightning/testnet",
+        \\ "our_features":{"init":"08a080282269a2","node":"88a080282269a2","channel":"","invoice":"02000020024100"}}}
     ;
     _ = try parser.parse(string);
 }

--- a/src/main.zig
+++ b/src/main.zig
@@ -33,7 +33,6 @@ const Mock = struct {
 };
 
 test "try to call core lightning Unix RPC method" {
-    //    const testing = @import("std").testing;
     const os = @import("std").os;
 
     const unix_path = os.getenv("CLN_UNIX") orelse unreachable;
@@ -41,6 +40,5 @@ test "try to call core lightning Unix RPC method" {
         return;
     };
     var allocator = std.heap.ArenaAllocator.init(std.heap.page_allocator);
-    // Looks like that we was not able to read all the json the the correct way
     _ = try client.call("getinfo", json.ObjectMap.init(allocator.allocator()));
 }

